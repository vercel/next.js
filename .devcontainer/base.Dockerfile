# syntax=docker.io/docker/dockerfile:1

# [Choice] Node.js version (use -bullseye variants on local arm64/Apple Silicon): 18-bullseye, 16-bullseye, 14-bullseye, 18-buster, 16-buster, 14-buster
ARG VARIANT=20-bullseye
FROM node:${VARIANT}

# [Option] Install zsh
ARG INSTALL_ZSH="true"
# [Option] Upgrade OS packages to their latest versions
ARG UPGRADE_PACKAGES="true"

# Install needed packages, yarn, nvm and setup non-root user. Use a separate RUN statement to add your own dependencies.
ARG USERNAME=node
ARG USER_UID=1000
ARG USER_GID=$USER_UID
ARG NPM_GLOBAL=/usr/local/share/npm-global
ENV NVM_DIR=/usr/local/share/nvm
ENV NVM_SYMLINK_CURRENT=true \ 
    PATH=${NPM_GLOBAL}/bin:${NVM_DIR}/current/bin:${PATH}
COPY library-scripts/*.sh library-scripts/*.env /tmp/library-scripts/
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive  \
    # Remove imagemagick due to https://security-tracker.debian.org/tracker/CVE-2019-10131
    && apt-get purge -y imagemagick imagemagick-6-common \
    # Install common packages, non-root user, update yarn and install nvm
    && bash /tmp/library-scripts/common-debian.sh "${INSTALL_ZSH}" "${USERNAME}" "${USER_UID}" "${USER_GID}" "${UPGRADE_PACKAGES}" "true" "true" \
    # Install yarn, nvm
    && rm -rf /opt/yarn-* /usr/local/bin/yarn /usr/local/bin/yarnpkg \
    && bash /tmp/library-scripts/node-debian.sh "${NVM_DIR}" "none" "${USERNAME}" \
    # Configure global npm install location, use group to adapt to UID/GID changes
    && if ! cat /etc/group | grep -e "^npm:" > /dev/null 2>&1; then groupadd -r npm; fi \
    && usermod -a -G npm ${USERNAME} \
    && umask 0002 \
    && mkdir -p ${NPM_GLOBAL} \
    && touch /usr/local/etc/npmrc \
    && chown ${USERNAME}:npm ${NPM_GLOBAL} /usr/local/etc/npmrc \
    && chmod g+s ${NPM_GLOBAL} \
    && npm config -g set prefix ${NPM_GLOBAL} \
    && sudo -u ${USERNAME} npm config -g set prefix ${NPM_GLOBAL} \
    # Install eslint
    && su ${USERNAME} -c "umask 0002 && npm install -g eslint" \
    && npm cache clean --force > /dev/null 2>&1 \
    # Install python-is-python3 on bullseye to prevent node-gyp regressions
    && . /etc/os-release \
    && if [ "${VERSION_CODENAME}" = "bullseye" ]; then apt-get -y install --no-install-recommends python-is-python3; fi \
    # Clean up
    && apt-get autoremove -y && apt-get clean -y && rm -rf /var/lib/apt/lists/* /root/.gnupg /tmp/library-scripts

# [Optional] Uncomment this section to install additional OS packages.
# RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
#     && apt-get -y install --no-install-recommends <your-package-list-here>

# [Optional] Uncomment if you want to install an additional version of node using nvm
# ARG EXTRA_NODE_VERSION=10
# RUN su node -c "source /usr/local/share/nvm/nvm.sh && nvm install ${EXTRA_NODE_VERSION}"

# [Optional] Uncomment if you want to install more global node modules
# RUN su node -c "npm install -g <your-package-list-here>""