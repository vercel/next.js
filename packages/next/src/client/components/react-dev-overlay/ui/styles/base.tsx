import * as React from 'react'
import { css } from '../../utils/css'

export function Base() {
  return (
    <style>
      {css`
        :host {
          ${
            // Although the style applied to the shadow host is isolated,
            // the element that attached the shadow host (i.e. `nextjs-portal`)
            // is still affected by the parent's style (e.g. `body`). This may
            // occur style conflicts like `display: flex`, with other children
            // elements therefore give the shadow host an absolute position.
            'position: absolute;'
          }

          --size-gap-half: 4px;
          --size-gap: 8px;
          --size-gap-double: 16px;
          --size-gap-triple: 24px;
          --size-gap-quad: 32px;

          --size-font-11: 11px;
          --size-font-smaller: 12px;
          --size-font-small: 14px;
          --size-font: 16px;
          --size-font-big: 20px;
          --size-font-bigger: 24px;

          --color-font: #757575;
          --color-backdrop: rgba(250, 250, 250, 0.8);
          --color-border-shadow: rgba(0, 0, 0, 0.145);

          --color-title-color: #1f1f1f;
          --color-stack-notes: #777;

          --color-accents-1: #808080;
          --color-accents-2: #222222;
          --color-accents-3: #404040;

          --font-stack-monospace: '__nextjs-Geist Mono', 'Geist Mono',
            'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier,
            monospace;
          --font-stack-sans: '__nextjs-Geist', 'Geist', -apple-system,
            'Source Sans Pro', sans-serif;

          font-family: var(--font-stack-sans);

          /* TODO: Remove replaced ones. */
          --shadow-sm: 0 1px 2px 0 rgb(0 0 0 / 0.05);
          --shadow: 0 1px 3px 0 rgb(0 0 0 / 0.1),
            0 1px 2px -1px rgb(0 0 0 / 0.1);
          --shadow-md: 0 4px 6px -1px rgb(0 0 0 / 0.1),
            0 2px 4px -2px rgb(0 0 0 / 0.1);
          --shadow-lg: 0 10px 15px -3px rgb(0 0 0 / 0.1),
            0 4px 6px -4px rgb(0 0 0 / 0.1);
          --shadow-xl: 0 20px 25px -5px rgb(0 0 0 / 0.1),
            0 8px 10px -6px rgb(0 0 0 / 0.1);
          --shadow-2xl: 0 25px 50px -12px rgb(0 0 0 / 0.25);
          --shadow-inner: inset 0 2px 4px 0 rgb(0 0 0 / 0.05);
          --shadow-none: 0 0 #0000;

          --shadow-small: 0px 2px 2px rgba(0, 0, 0, 0.04);
          --shadow-menu: 0px 1px 1px rgba(0, 0, 0, 0.02),
            0px 4px 8px -4px rgba(0, 0, 0, 0.04),
            0px 16px 24px -8px rgba(0, 0, 0, 0.06);

          --focus-color: var(--color-blue-800);
          --focus-ring: 2px solid var(--focus-color);

          --timing-swift: cubic-bezier(0.23, 0.88, 0.26, 0.92);
          --timing-overlay: cubic-bezier(0.175, 0.885, 0.32, 1.1);

          --rounded-none: 0px;
          --rounded-sm: 0.125rem; /* 2px */
          --rounded-md: 0.25rem; /* 4px */
          --rounded-md-2: 0.375rem; /* 6px */
          --rounded-lg: 0.5rem; /* 8px */
          --rounded-xl: 0.75rem; /* 12px */
          --rounded-2xl: 1rem; /* 16px */
          --rounded-3xl: 1.5rem; /* 24px */
          --rounded-full: 9999px;

          --size-0: 0px;
          --size-px: 1px;
          --size-0_5: 0.125rem; /* 2px */
          --size-1: 0.25rem; /* 4px */
          --size-1_5: 0.375rem; /* 6px */
          --size-2: 0.5rem; /* 8px */
          --size-2_5: 0.625rem; /* 10px */
          --size-3: 0.75rem; /* 12px */
          --size-3_5: 0.875rem; /* 14px */
          --size-4: 1rem; /* 16px */
          --size-4_5: 1.125rem; /* 18px */
          --size-5: 1.25rem; /* 20px */
          --size-5_5: 1.375rem; /* 22px */
          --size-6: 1.5rem; /* 24px */
          --size-6_5: 1.625rem; /* 26px */
          --size-7: 1.75rem; /* 28px */
          --size-7_5: 1.875rem; /* 30px */
          --size-8: 2rem; /* 32px */
          --size-8_5: 2.125rem; /* 34px */
          --size-9: 2.25rem; /* 36px */
          --size-9_5: 2.375rem; /* 38px */
          --size-10: 2.5rem; /* 40px */
          --size-10_5: 2.625rem; /* 42px */
          --size-11: 2.75rem; /* 44px */
          --size-11_5: 2.875rem; /* 46px */
          --size-12: 3rem; /* 48px */

          @media print {
            display: none;
          }
        }

        @media (prefers-color-scheme: dark) {
          :host {
            --color-font: white;
            --color-backdrop: rgb(0, 0, 0, 0.8);
            --color-border-shadow: rgba(255, 255, 255, 0.145);

            --color-title-color: #fafafa;
            --color-stack-notes: #a9a9a9;
          }
        }

        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
          margin-bottom: var(--size-gap);
          font-weight: 500;
          line-height: 1.5;
        }

        a {
          &:focus {
            outline: var(--focus-ring);
          }
        }
      `}
    </style>
  )
}
