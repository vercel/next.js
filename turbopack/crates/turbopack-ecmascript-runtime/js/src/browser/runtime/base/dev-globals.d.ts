/**
 * Definitions for globals that are injected by the Turbopack runtime.
 *
 * These are available from every module, but should only be used by Turbopack
 * code, not by user code.
 */

type UpdateCallback = (update: ServerMessage) => void;

type ChunkUpdateProvider = {
  push: (registration: [ChunkPath, UpdateCallback]) => void;
};

declare var TURBOPACK_CHUNK_UPDATE_LISTENERS:
  | ChunkUpdateProvider
  | [ChunkPath, UpdateCallback][]
  | undefined;
// This is used by the Next.js integration test suite to notify it when HMR
// updates have been completed.
declare var __NEXT_HMR_CB: undefined | null | (() => void);
