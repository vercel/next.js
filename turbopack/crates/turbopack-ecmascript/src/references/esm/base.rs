use anyhow::{anyhow, bail, Result};
use strsim::jaro;
use swc_core::{
    common::{BytePos, Span, DUMMY_SP},
    ecma::ast::{Decl, Expr, ExprStmt, Ident, Stmt},
    quote,
};
use turbo_rcstr::RcStr;
use turbo_tasks::{ResolvedVc, Value, ValueToString, Vc};
use turbo_tasks_fs::FileSystemPath;
use turbopack_core::{
    chunk::{
        ChunkItemExt, ChunkableModule, ChunkableModuleReference, ChunkingContext, ChunkingType,
        ChunkingTypeOption,
    },
    issue::{
        Issue, IssueExt, IssueSeverity, IssueSource, IssueStage, OptionIssueSource,
        OptionStyledString, StyledString,
    },
    module::Module,
    reference::ModuleReference,
    reference_type::{EcmaScriptModulesReferenceSubType, ImportWithType},
    resolve::{
        origin::{ResolveOrigin, ResolveOriginExt},
        parse::Request,
        ExternalType, ModulePart, ModuleResolveResult, ModuleResolveResultItem,
    },
};
use turbopack_resolve::ecmascript::esm_resolve;

use super::export::{all_known_export_names, is_export_missing};
use crate::{
    analyzer::imports::ImportAnnotations,
    chunk::EcmascriptChunkPlaceable,
    code_gen::{CodeGenerateable, CodeGeneration},
    magic_identifier,
    references::util::{request_to_string, throw_module_not_found_expr},
    tree_shake::{asset::EcmascriptModulePartAsset, TURBOPACK_PART_IMPORT_SOURCE},
    utils::module_id_to_lit,
};

#[turbo_tasks::value]
pub enum ReferencedAsset {
    Some(ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>),
    External(RcStr, ExternalType),
    None,
    Unresolvable,
}

impl ReferencedAsset {
    pub async fn get_ident(&self) -> Result<Option<String>> {
        Ok(match self {
            ReferencedAsset::Some(asset) => Some(Self::get_ident_from_placeable(asset).await?),
            ReferencedAsset::External(request, ty) => Some(magic_identifier::mangle(&format!(
                "{ty} external {request}"
            ))),
            ReferencedAsset::None | ReferencedAsset::Unresolvable => None,
        })
    }

    pub(crate) async fn get_ident_from_placeable(
        asset: &Vc<Box<dyn EcmascriptChunkPlaceable>>,
    ) -> Result<String> {
        let path = asset.ident().to_string().await?;
        Ok(magic_identifier::mangle(&format!(
            "imported module {}",
            path
        )))
    }
}

#[turbo_tasks::value_impl]
impl ReferencedAsset {
    #[turbo_tasks::function]
    pub async fn from_resolve_result(resolve_result: Vc<ModuleResolveResult>) -> Result<Vc<Self>> {
        // TODO handle multiple keyed results
        let result = resolve_result.await?;
        if result.is_unresolvable_ref() {
            return Ok(ReferencedAsset::Unresolvable.cell());
        }
        for (_, result) in result.primary.iter() {
            match result {
                ModuleResolveResultItem::External {
                    name: request, ty, ..
                } => {
                    return Ok(ReferencedAsset::External(request.clone(), *ty).cell());
                }
                &ModuleResolveResultItem::Module(module) => {
                    if let Some(placeable) =
                        ResolvedVc::try_downcast::<Box<dyn EcmascriptChunkPlaceable>>(module)
                            .await?
                    {
                        return Ok(ReferencedAsset::Some(placeable).cell());
                    }
                }
                // TODO ignore should probably be handled differently
                _ => {}
            }
        }
        Ok(ReferencedAsset::None.cell())
    }
}

#[turbo_tasks::value]
#[derive(Hash, Debug)]
pub struct EsmAssetReference {
    pub origin: ResolvedVc<Box<dyn ResolveOrigin>>,
    pub request: ResolvedVc<Request>,
    pub annotations: ImportAnnotations,
    pub issue_source: ResolvedVc<IssueSource>,
    pub export_name: Option<ResolvedVc<ModulePart>>,
    pub import_externals: bool,
}

impl EsmAssetReference {
    fn get_origin(&self) -> Vc<Box<dyn ResolveOrigin>> {
        if let Some(transition) = self.annotations.transition() {
            self.origin.with_transition(transition.into())
        } else {
            *self.origin
        }
    }
}

#[turbo_tasks::value_impl]
impl EsmAssetReference {
    #[turbo_tasks::function]
    pub fn new(
        origin: ResolvedVc<Box<dyn ResolveOrigin>>,
        request: ResolvedVc<Request>,
        issue_source: ResolvedVc<IssueSource>,
        annotations: Value<ImportAnnotations>,
        export_name: Option<ResolvedVc<ModulePart>>,
        import_externals: bool,
    ) -> Vc<Self> {
        Self::cell(EsmAssetReference {
            origin,
            request,
            issue_source,
            annotations: annotations.into_value(),
            export_name,
            import_externals,
        })
    }

    #[turbo_tasks::function]
    pub(crate) fn get_referenced_asset(self: Vc<Self>) -> Vc<ReferencedAsset> {
        ReferencedAsset::from_resolve_result(self.resolve_reference())
    }
}

#[turbo_tasks::value_impl]
impl ModuleReference for EsmAssetReference {
    #[turbo_tasks::function]
    async fn resolve_reference(&self) -> Result<Vc<ModuleResolveResult>> {
        let ty = if matches!(self.annotations.module_type(), Some("json")) {
            EcmaScriptModulesReferenceSubType::ImportWithType(ImportWithType::Json)
        } else if let Some(part) = &self.export_name {
            EcmaScriptModulesReferenceSubType::ImportPart(*part)
        } else {
            EcmaScriptModulesReferenceSubType::Import
        };

        if let Request::Module { module, .. } = &*self.request.await? {
            if module == TURBOPACK_PART_IMPORT_SOURCE {
                if let Some(part) = self.export_name {
                    let module: ResolvedVc<crate::EcmascriptModuleAsset> =
                        ResolvedVc::try_downcast_type(self.origin)
                            .await?
                            .expect("EsmAssetReference origin should be a EcmascriptModuleAsset");

                    return Ok(ModuleResolveResult::module(
                        EcmascriptModulePartAsset::select_part(*module, *part)
                            .to_resolved()
                            .await?,
                    )
                    .cell());
                }

                bail!("export_name is required for part import")
            }
        }

        let result = esm_resolve(
            self.get_origin().resolve().await?,
            *self.request,
            Value::new(ty),
            false,
            Some(*self.issue_source),
        );

        if let Some(part) = self.export_name {
            let part = part.await?;
            if let &ModulePart::Export(export_name) = &*part {
                for &module in result.primary_modules().await? {
                    if let Some(module) = ResolvedVc::try_downcast(module).await? {
                        let export = export_name.await?;
                        if *is_export_missing(*module, export.clone_value()).await? {
                            InvalidExport {
                                export: export_name,
                                module,
                                source: self.issue_source,
                            }
                            .resolved_cell()
                            .emit();
                        }
                    }
                }
            }
        }

        Ok(result)
    }
}

#[turbo_tasks::value_impl]
impl ValueToString for EsmAssetReference {
    #[turbo_tasks::function]
    async fn to_string(&self) -> Result<Vc<RcStr>> {
        Ok(Vc::cell(
            format!(
                "import {} with {}",
                self.request.to_string().await?,
                self.annotations
            )
            .into(),
        ))
    }
}

#[turbo_tasks::value_impl]
impl ChunkableModuleReference for EsmAssetReference {
    #[turbo_tasks::function]
    fn chunking_type(&self) -> Result<Vc<ChunkingTypeOption>> {
        Ok(Vc::cell(
            if let Some(chunking_type) = self.annotations.chunking_type() {
                match chunking_type {
                    "parallel" => Some(ChunkingType::ParallelInheritAsync),
                    "none" => None,
                    _ => return Err(anyhow!("unknown chunking_type: {}", chunking_type)),
                }
            } else {
                Some(ChunkingType::ParallelInheritAsync)
            },
        ))
    }
}

#[turbo_tasks::value_impl]
impl CodeGenerateable for EsmAssetReference {
    #[turbo_tasks::function]
    async fn code_generation(
        self: Vc<Self>,
        chunking_context: Vc<Box<dyn ChunkingContext>>,
    ) -> Result<Vc<CodeGeneration>> {
        let this = &*self.await?;

        // only chunked references can be imported
        let result = if this.annotations.chunking_type() != Some("none") {
            let import_externals = this.import_externals;
            let referenced_asset = self.get_referenced_asset().await?;
            if let ReferencedAsset::Unresolvable = &*referenced_asset {
                // Insert code that throws immediately at time of import if a request is
                // unresolvable
                let request = request_to_string(*this.request).await?.to_string();
                let stmt = Stmt::Expr(ExprStmt {
                    expr: Box::new(throw_module_not_found_expr(&request)),
                    span: DUMMY_SP,
                });
                Some((format!("throw {request}").into(), stmt))
            } else if let Some(ident) = referenced_asset.get_ident().await? {
                let span = this
                    .issue_source
                    .await?
                    .to_swc_offsets()
                    .await?
                    .map_or(DUMMY_SP, |(start, end)| {
                        Span::new(BytePos(start as u32), BytePos(end as u32))
                    });
                match &*referenced_asset {
                    ReferencedAsset::Unresolvable => {
                        unreachable!()
                    }
                    ReferencedAsset::Some(asset) => {
                        let id = asset
                            .as_chunk_item(Vc::upcast(chunking_context))
                            .id()
                            .await?;
                        Some((
                            ident.clone().into(),
                            var_decl_with_span(
                                quote!(
                                    "var $name = __turbopack_import__($id);" as Stmt,
                                    name = Ident::new(ident.clone().into(), DUMMY_SP, Default::default()),
                                    id: Expr = module_id_to_lit(&id),
                                ),
                                span,
                            ),
                        ))
                    }
                    ReferencedAsset::External(request, ExternalType::EcmaScriptModule) => {
                        if !*chunking_context
                            .environment()
                            .supports_esm_externals()
                            .await?
                        {
                            bail!(
                                "the chunking context ({}) does not support external modules (esm \
                                 request: {})",
                                chunking_context.name().await?,
                                request
                            );
                        }
                        Some((
                            ident.clone().into(),
                            var_decl_with_span(
                                if import_externals {
                                    quote!(
                                        "var $name = __turbopack_external_import__($id);" as Stmt,
                                        name = Ident::new(ident.clone().into(), DUMMY_SP, Default::default()),
                                        id: Expr = Expr::Lit(request.clone().to_string().into())
                                    )
                                } else {
                                    quote!(
                                        "var $name = __turbopack_external_require__($id, () => require($id), true);" as Stmt,
                                        name = Ident::new(ident.clone().into(), DUMMY_SP, Default::default()),
                                        id: Expr = Expr::Lit(request.clone().to_string().into())
                                    )
                                },
                                span,
                            ),
                        ))
                    }
                    ReferencedAsset::External(
                        request,
                        ExternalType::CommonJs | ExternalType::Url,
                    ) => {
                        if !*chunking_context
                            .environment()
                            .supports_commonjs_externals()
                            .await?
                        {
                            bail!(
                                "the chunking context ({}) does not support external modules \
                                 (request: {})",
                                chunking_context.name().await?,
                                request
                            );
                        }
                        Some((
                            ident.clone().into(),
                            var_decl_with_span(
                                quote!(
                                    "var $name = __turbopack_external_require__($id, () => require($id), true);" as Stmt,
                                    name = Ident::new(ident.clone().into(), DUMMY_SP, Default::default()),
                                    id: Expr = Expr::Lit(request.clone().to_string().into())
                                ),
                                span,
                            ),
                        ))
                    }
                    // fallback in case we introduce a new `ExternalType`
                    #[allow(unreachable_patterns)]
                    ReferencedAsset::External(request, ty) => {
                        bail!(
                            "Unsupported external type {:?} for ESM reference with request: {:?}",
                            ty,
                            request
                        )
                    }
                    ReferencedAsset::None => None,
                }
            } else {
                None
            }
        } else {
            None
        };

        if let Some((key, stmt)) = result {
            Ok(CodeGeneration::hoisted_stmt(key, stmt))
        } else {
            Ok(CodeGeneration::empty())
        }
    }
}

fn var_decl_with_span(mut decl: Stmt, span: Span) -> Stmt {
    match &mut decl {
        Stmt::Decl(Decl::Var(decl)) => decl.span = span,
        _ => panic!("Expected Stmt::Decl::Var"),
    };
    decl
}

#[turbo_tasks::value(shared)]
pub struct InvalidExport {
    export: ResolvedVc<RcStr>,
    module: ResolvedVc<Box<dyn EcmascriptChunkPlaceable>>,
    source: ResolvedVc<IssueSource>,
}

#[turbo_tasks::value_impl]
impl Issue for InvalidExport {
    #[turbo_tasks::function]
    fn severity(&self) -> Vc<IssueSeverity> {
        IssueSeverity::Error.into()
    }

    #[turbo_tasks::function]
    async fn title(&self) -> Result<Vc<StyledString>> {
        Ok(StyledString::Line(vec![
            StyledString::Text("Export ".into()),
            StyledString::Code(self.export.await?.clone_value()),
            StyledString::Text(" doesn't exist in target module".into()),
        ])
        .cell())
    }

    #[turbo_tasks::function]
    fn stage(&self) -> Vc<IssueStage> {
        IssueStage::Bindings.into()
    }

    #[turbo_tasks::function]
    fn file_path(&self) -> Vc<FileSystemPath> {
        self.source.file_path()
    }

    #[turbo_tasks::function]
    async fn description(&self) -> Result<Vc<OptionStyledString>> {
        let export = self.export.await?;
        let export_names = all_known_export_names(*self.module).await?;
        let did_you_mean = export_names
            .iter()
            .map(|s| (s, jaro(export.as_str(), s.as_str())))
            .max_by(|a, b| a.1.partial_cmp(&b.1).unwrap())
            .map(|(s, _)| s);
        Ok(Vc::cell(Some(
            StyledString::Stack(vec![
                StyledString::Line(vec![
                    StyledString::Text("The export ".into()),
                    StyledString::Code(export.clone_value()),
                    StyledString::Text(" was not found in module ".into()),
                    StyledString::Strong(self.module.ident().to_string().await?.clone_value()),
                    StyledString::Text(".".into()),
                ]),
                if let Some(did_you_mean) = did_you_mean {
                    StyledString::Line(vec![
                        StyledString::Text("Did you mean to import ".into()),
                        StyledString::Code(did_you_mean.clone()),
                        StyledString::Text("?".into()),
                    ])
                } else {
                    StyledString::Strong("The module has no exports at all.".into())
                },
                StyledString::Text(
                    "All exports of the module are statically known (It doesn't have dynamic \
                     exports). So it's known statically that the requested export doesn't exist."
                        .into(),
                ),
            ])
            .resolved_cell(),
        )))
    }

    #[turbo_tasks::function]
    async fn detail(&self) -> Result<Vc<OptionStyledString>> {
        let export_names = all_known_export_names(*self.module).await?;
        Ok(Vc::cell(Some(
            StyledString::Line(vec![
                StyledString::Text("These are the exports of the module:\n".into()),
                StyledString::Code(
                    export_names
                        .iter()
                        .map(|s| s.as_str())
                        .intersperse(", ")
                        .collect::<String>()
                        .into(),
                ),
            ])
            .resolved_cell(),
        )))
    }

    #[turbo_tasks::function]
    fn source(&self) -> Vc<OptionIssueSource> {
        Vc::cell(Some(self.source))
    }
}
