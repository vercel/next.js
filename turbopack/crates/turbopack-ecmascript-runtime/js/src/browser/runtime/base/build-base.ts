/// <reference path="./runtime-base.ts" />
/// <reference path="./dummy.ts" />

declare var augmentContext: ((context: unknown) => unknown);

const moduleCache: ModuleCache<Module> = {};

/**
 * Gets or instantiates a runtime module.
 */
// @ts-ignore
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function getOrInstantiateRuntimeModule(
  moduleId: ModuleId,
  chunkPath: ChunkPath,
): Module {
  const module = moduleCache[moduleId];
  if (module) {
    if (module.error) {
      throw module.error;
    }
    return module;
  }

  return instantiateModule(moduleId, { type: SourceType.Runtime, chunkPath });
}

/**
 * Retrieves a module from the cache, or instantiate it if it is not cached.
 */
// Used by the backend
// @ts-ignore
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const getOrInstantiateModuleFromParent: GetOrInstantiateModuleFromParent<Module> = (
  id,
  sourceModule
) => {
  const module = moduleCache[id];

  if (module) {
    return module;
  }

  return instantiateModule(id, {
    type: SourceType.Parent,
    parentId: sourceModule.id,
  });
};

function instantiateModule(id: ModuleId, source: SourceInfo): Module {
  const moduleFactory = moduleFactories[id];
  if (typeof moduleFactory !== "function") {
    // This can happen if modules incorrectly handle HMR disposes/updates,
    // e.g. when they keep a `setTimeout` around which still executes old code
    // and contains e.g. a `require("something")` call.
    let instantiationReason;
    switch (source.type) {
      case SourceType.Runtime:
        instantiationReason = `as a runtime entry of chunk ${source.chunkPath}`;
        break;
      case SourceType.Parent:
        instantiationReason = `because it was required from module ${source.parentId}`;
        break;
      case SourceType.Update:
        instantiationReason = "because of an HMR update";
        break;
      default:
        invariant(source, (source) => `Unknown source type: ${source?.type}`);
    }
    throw new Error(
      `Module ${id} was instantiated ${instantiationReason}, but the module factory is not available. It might have been deleted in an HMR update.`
    );
  }

  switch (source.type) {
    case SourceType.Runtime:
      runtimeModules.add(id);
      break;
    case SourceType.Parent:
      // No need to add this module as a child of the parent module here, this
      // has already been taken care of in `getOrInstantiateModuleFromParent`.
      break;
    case SourceType.Update:
      throw new Error('Unexpected')
    default:
      invariant(source, (source) => `Unknown source type: ${source?.type}`);
  }

  const module: Module = {
    exports: {},
    error: undefined,
    loaded: false,
    id,
    namespaceObject: undefined,
  };

  moduleCache[id] = module;

  // NOTE(alexkirsz) This can fail when the module encounters a runtime error.
  try {
    const sourceInfo: SourceInfo = { type: SourceType.Parent, parentId: id };

    const r = commonJsRequire.bind(null, module);
    moduleFactory.call(
      module.exports,
      augmentContext({
        a: asyncModule.bind(null, module),
        e: module.exports,
        r: commonJsRequire.bind(null, module),
        t: runtimeRequire,
        f: moduleContext,
        i: esmImport.bind(null, module),
        s: esmExport.bind(null, module, module.exports),
        j: dynamicExport.bind(null, module, module.exports),
        v: exportValue.bind(null, module),
        n: exportNamespace.bind(null, module),
        m: module,
        c: moduleCache,
        M: moduleFactories,
        l: loadChunk.bind(null, sourceInfo),
        w: loadWebAssembly.bind(null, sourceInfo),
        u: loadWebAssemblyModule.bind(null, sourceInfo),
        g: globalThis,
        P: resolveAbsolutePath,
        U: relativeURL,
        R: createResolvePathFromModule(r),
        b: getWorkerBlobURL,
        __dirname: typeof module.id === "string" ? module.id.replace(/(^|\/)\/+$/, "") : module.id
      })
    );
  } catch (error) {
    module.error = error as any;
    throw error;
  }

  module.loaded = true;
  if (module.namespaceObject && module.exports !== module.namespaceObject) {
    // in case of a circular dependency: cjs1 -> esm2 -> cjs1
    interopEsm(module.exports, module.namespaceObject);
  }

  return module;
}

