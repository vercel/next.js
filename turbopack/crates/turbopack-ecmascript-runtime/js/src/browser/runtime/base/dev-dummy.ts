/**
 * This file acts as a dummy implementor for the interface that
 * `runtime-base.ts` expects to be available in the global scope.
 *
 * This interface will be implemented by runtime backends.
 */

/* eslint-disable @typescript-eslint/no-unused-vars */

declare var DEV_BACKEND: DevRuntimeBackend;
declare var _eval: (code: EcmascriptModuleEntry) => any;
/**
 * Adds additional properties to the `TurbopackDevBaseContext` interface.
 */
declare var augmentContext: (
  context: TurbopackDevBaseContext
) => TurbopackDevContext;
