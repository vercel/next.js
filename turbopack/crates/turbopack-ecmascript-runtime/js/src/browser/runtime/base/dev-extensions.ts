/// <reference path="../../../shared/runtime-utils.ts" />

/**
 * Extensions to the shared runtime types that are specific to the development
 * runtime (e.g. `module.hot`).
 */

declare const enum HotUpdateStatus {
  idle = "idle",
}

type HotUpdateStatusHandler = (status: HotUpdateStatus) => void;

interface HotData {
  prevExports?: Exports;
}

// Used through reference comments
// eslint-disable-next-line @typescript-eslint/no-unused-vars
interface HotState {
  selfAccepted: boolean | Function;
  selfDeclined: boolean;
  selfInvalidated: boolean;
  disposeHandlers: ((data: object) => void)[];
}

type AcceptErrorHandler = (
  err: Error,
  context: { moduleId: ModuleId; dependencyId: string | number }
) => void;
type AcceptCallback = (outdatedDependencies: string[]) => void;

interface AcceptFunction {
  // accept updates for self
  (errorHandler?: AcceptErrorHandler): void;

  // accept updates for the given modules
  (
    modules?: string | string[],
    callback?: AcceptCallback,
    errorHandler?: AcceptErrorHandler
  ): void;
}

interface Hot {
  active: boolean;
  data: HotData;

  accept: AcceptFunction;

  decline: (module?: string | string[]) => void;

  dispose: (callback: (data: HotData) => void) => void;

  addDisposeHandler: (callback: (data: object) => void) => void;

  removeDisposeHandler: (callback: (data: object) => void) => void;

  invalidate: () => void;

  status: () => keyof typeof HotUpdateStatus;
  addStatusHandler: (handler: HotUpdateStatusHandler) => void;
  removeStatusHandler: (handler: HotUpdateStatusHandler) => void;
  check: (autoApply: boolean) => Promise<any[] | null>;
}

// Used through reference comments
// eslint-disable-next-line @typescript-eslint/no-unused-vars
interface HotModule extends ModuleWithDirection {
  hot: Hot;
}
