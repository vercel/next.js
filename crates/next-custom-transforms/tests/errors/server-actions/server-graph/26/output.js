import { BaseClass } from './base';
export class MyClass extends BaseClass {
    static async foo() {
        // super is allowed here
        super.foo();
        return fetch('https://example.com').then((res)=>res.json());
    }
    static async bar() {
        if (Math.random() > 0.5) {
            // super is not allowed here
            return [
                super.bar()
            ];
        }
        // arguments is not allowed here
        console.log(arguments);
        // this is not allowed here
        return this.foo();
    }
}
