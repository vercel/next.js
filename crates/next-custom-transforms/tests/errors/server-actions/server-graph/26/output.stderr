  x "use cache" functions cannot use `super`.
  | 
    ,-[input.js:15:1]
 14 |       // super is not allowed here
 15 |       return [super.bar()]
    :               ^^^^^
 16 |     }
    `----
  x "use cache" functions cannot use `arguments`.
  | 
    ,-[input.js:19:1]
 18 |     // arguments is not allowed here
 19 |     console.log(arguments)
    :                 ^^^^^^^^^
 20 |     // this is not allowed here
    `----
  x "use cache" functions cannot use `this`.
  | 
    ,-[input.js:21:1]
 20 |     // this is not allowed here
 21 |     return this.foo()
    :            ^^^^
 22 |   }
    `----
